rfcs/2023-04-18-bundle-check.md

```diff
@@ -0,0 +1,124 @@
+
+# Motivation
+As Blade grows, the size of the generated bundles also grows. This can lead to longer load times and increased bandwidth usage for our consumers. To prevent this, we need a way to monitor the size of our generated assets and ensure that they stay within acceptable limits. We also need to ensure that any changes we make to the design system don't inadvertently increase the size of our bundles, which can impact our users' experience.
+
+# Detailed Design
+We want to achieve the following:
+1. Add a check for each PR that ensures size of our bundles don't increase beyond a certain limit.
+2. Enable a way for the PR author to visualize the impact of their changes on the bundle size by having a diff from master's bundle size.
+3. Ensure our tree-shaking works as expected by adding a check to ensure that unused components are not included in the bundle.
+
+### Bundle Size Limit & Master Diff
+We will achieve the following with this:
+1. Add a check for each PR that ensures size of our bundles don't increase beyond a certain limit.
+2. Enable a way for the PR author to visualize the impact of their changes on the bundle size by having a diff from master's bundle size.
+
+#### Bundlewatch
+- We will use [bundlewatch](bundlewatch.io) to achieve this. Bundlewatch is a tool that helps us to keep an eye on our bundle size. It works by comparing the size of our bundles against a predefined limit. If the size of the bundle exceeds the limit, the build fails. It also provides a diff between the current bundle size and the size of the bundle on master. This helps us to visualize the impact of our changes on the bundle size.
+- Bundlewatch allows us to pass a config like this in `package.json`:
+```json
+{
+  "files": [
+    {
+      "path": "./build/components/index.web.js",
+      "maxSize": "200kB"
+    },
+    {
+      "path": "./build/tokens/index.web.js",
+      "maxSize": "100kB"
+    },
+    {
+      "path": "./build/components/index.native.js",
+      "maxSize": "200kB"
+    },
+    {
+      "path": "./build/tokens/index.native.js",
+      "maxSize": "100kB"
+    }
+  ]
+}
+```
+- We will need to add a github action to run bundlewatch on every PR. This can be a part of `blade-validate`(for PR) + `release`(for master push) workflow or a separate `bundle-check` workflow(that triggers on PR & master push).
+- This github action will pass or fail based on the limits defined in the config.
+- To get a diff of master we will need to allow Bundlewatch to add a build-status check to the PR (requires github admin approval). The status check will show the diff between the current bundle size and the size of the bundle on master. This will help us to visualize the impact of our changes on the bundle size.
+- The build-status check will show the bundle size like this:
+<img src="./images/bundle-size/build-status.png" width="800" />
+
+- When you click on the `Details` link, it will take you to a page like this:
```

Comment:
Can we see components wise breakdown? For example this PR increased bundle size for Button by x and increased bundle size for Text by y?

Reply:
Couldn't find any good solutions for this, we could use `size-limit` and add limits for individual components. It won't do diff but it will at least ensure a specific component doesn't grow in size. I will continue exploring more solutions parallely to see if we can get component diff. This is a good starting point though